snow/engine/avalanche/vertex/stateless_vertex.go

```diff
@@ -0,0 +1,108 @@
+
+package vertex
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/components/verify"
+)
+
+const (
+	// maxNumParents is the max number of parents a vertex may have
+	maxNumParents = 128
+
+	// maxTxsPerVtx is the max number of transactions a vertex may have
+	maxTxsPerVtx = 128
+)
+
+var (
+	errBadVersion          = errors.New("invalid version")
+	errBadEpoch            = errors.New("invalid epoch")
+	errFutureField         = errors.New("field specified in a previous version")
+	errTooManyparentIDs    = fmt.Errorf("vertex contains more than %d parentIDs", maxNumParents)
+	errNoOperations        = errors.New("vertex contains no operations")
+	errTooManyTxs          = fmt.Errorf("vertex contains more than %d transactions", maxTxsPerVtx)
+	errTooManyRestrictions = fmt.Errorf("vertex contains more than %d restrictions", maxTxsPerVtx)
+	errInvalidParents      = errors.New("vertex contains non-sorted or duplicated parentIDs")
+	errInvalidRestrictions = errors.New("vertex contains non-sorted or duplicated restrictions")
+	errInvalidTxs          = errors.New("vertex contains non-sorted or duplicated transactions")
+
+	_ StatelessVertex = statelessVertex{}
+)
+
+type StatelessVertex interface {
+	verify.Verifiable
+	ID() ids.ID
+	Bytes() []byte
+
+	Version() uint16
+	ChainID() ids.ID
+	Height() uint64
+	Epoch() uint32
+	ParentIDs() []ids.ID
+	Txs() [][]byte
+	Restrictions() []ids.ID
+}
+
+type statelessVertex struct {
+	// This wrapper exists so that the function calls aren't ambiguous
+	innerStatelessVertex
+
+	// cache the ID of this vertex
+	id ids.ID
+
+	// cache the binary format of this vertex
+	bytes []byte
+}
+
+func (v statelessVertex) ID() ids.ID             { return v.id }
+func (v statelessVertex) Bytes() []byte          { return v.bytes }
+func (v statelessVertex) Version() uint16        { return v.innerStatelessVertex.Version }
+func (v statelessVertex) ChainID() ids.ID        { return v.innerStatelessVertex.ChainID }
+func (v statelessVertex) Height() uint64         { return v.innerStatelessVertex.Height }
+func (v statelessVertex) Epoch() uint32          { return v.innerStatelessVertex.Epoch }
+func (v statelessVertex) ParentIDs() []ids.ID    { return v.innerStatelessVertex.ParentIDs }
+func (v statelessVertex) Txs() [][]byte          { return v.innerStatelessVertex.Txs }
+func (v statelessVertex) Restrictions() []ids.ID { return v.innerStatelessVertex.Restrictions }
+
+type innerStatelessVertex struct {
+	Version      uint16   `json:"version"`
+	ChainID      ids.ID   `serializeV0:"true" serializeV1:"true" json:"chainID"`
+	Height       uint64   `serializeV0:"true" serializeV1:"true" json:"height"`
+	Epoch        uint32   `serializeV0:"true" serializeV1:"true" json:"epoch"`
+	ParentIDs    []ids.ID `serializeV0:"true" serializeV1:"true" len:"128" json:"parentIDs"`
+	Txs          [][]byte `serializeV0:"true" serializeV1:"true" len:"128" json:"txs"`
```

Comment:
Should this field be the transaction itself instead of the byte representation, so that there's no need to recompute its hash when sorting?

Reply:
What do you mean by "the transaction itself"? This struct shouldn't have any references to persistent storage. Which having a reference to a `snowstorm.Tx` is too many dependencies.